scene/main/window.cpp

```diff
@@ -672,16 +672,21 @@ void Window::_event_callback(DisplayServer::WindowEvent p_event) {
 		case DisplayServer::WINDOW_EVENT_MOUSE_ENTER: {
 			_propagate_window_notification(this, NOTIFICATION_WM_MOUSE_ENTER);
 			Window *root = get_tree()->get_root();
-			DEV_ASSERT(!root->gui.windowmanager_window_over); // Entering a window while a window is hovered should never happen.
+			Window *previous_over = root->gui.windowmanager_window_over;
+			if (previous_over) {
+				ERR_FAIL_COND_MSG(previous_over == this, "Entering a window while hovering it should never happen.");
+				previous_over->_event_callback(DisplayServer::WINDOW_EVENT_MOUSE_EXIT);
+			}
 			root->gui.windowmanager_window_over = this;
 			notification(NOTIFICATION_VP_MOUSE_ENTER);
 			if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_CURSOR_SHAPE)) {
 				DisplayServer::get_singleton()->cursor_set_shape(DisplayServer::CURSOR_ARROW); //restore cursor shape
 			}
+			ERR_FAIL_COND_MSG(previous_over, "Entering a window while a window is hovered should never happen.");
 		} break;
 		case DisplayServer::WINDOW_EVENT_MOUSE_EXIT: {
 			Window *root = get_tree()->get_root();
-			DEV_ASSERT(root->gui.windowmanager_window_over); // Exiting a window, while no window is hovered should never happen.
+			ERR_FAIL_NULL_MSG(root->gui.windowmanager_window_over, "Exiting a window, while no window is hovered should never happen.");
```

Comment:
```suggestion
			ERR_FAIL_NULL_MSG(root->gui.windowmanager_window_over, "Exiting a window while no window is hovered should never happen.");
```

But "should never happen" is still the kind of message that one would use for a `DEV_ASSERT`.

If it can be triggered by user code, and thus require an `ERR_FAIL` for graceful handling, then it does actually happen and the message should clarify what went wrong so that the user can fix it on their side.

Reply:
Though I guess here the "user" can be a platform port that does things wrong in its window handling? So I guess it makes sense as is.